core/src/apps/cardano/sign_tx.py

```diff
@@ -67,24 +71,70 @@
 async def sign_tx(
     ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
 ) -> CardanoSignedTx:
-    try:
-        if msg.fee > LOVELACE_MAX_SUPPLY:
-            raise wire.ProcessError("Fee is out of range!")
+    if msg.fee > LOVELACE_MAX_SUPPLY:
+        raise wire.ProcessError("Fee is out of range!")
+
+    validate_network_info(msg.network_id, msg.protocol_magic)
+
+    if _has_stake_pool_registration(msg):
+        return await _sign_stake_pool_registration_tx(ctx, msg, keychain)
+    else:
+        return await _sign_standard_tx(ctx, msg, keychain)
 
-        validate_network_info(msg.network_id, msg.protocol_magic)
 
+async def _sign_standard_tx(
+    ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
+) -> CardanoSignedTx:
+    try:
         for i in msg.inputs:
             await validate_path(
                 ctx, keychain, i.address_n, SCHEMA_ADDRESS.match(i.address_n)
             )
 
         _validate_outputs(keychain, msg.outputs, msg.protocol_magic, msg.network_id)
-        _validate_certificates(msg.certificates)
+        _validate_certificates(msg.certificates, msg.protocol_magic, msg.network_id)
         _validate_withdrawals(msg.withdrawals)
         _validate_metadata(msg.metadata)
 
         # display the transaction in UI
-        await _show_tx(ctx, keychain, msg)
+        await _show_standard_tx(ctx, keychain, msg)
+
+        # sign the transaction bundle and prepare the result
+        serialized_tx, tx_hash = _serialize_tx(keychain, msg)
+        tx = CardanoSignedTx(serialized_tx=serialized_tx, tx_hash=tx_hash)
+
+    except ValueError as e:
+        if __debug__:
+            log.exception(__name__, e)
+        raise wire.ProcessError("Signing failed")
+
+    return tx
+
+
+async def _sign_stake_pool_registration_tx(
+    ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
+) -> CardanoSignedTx:
+    """
+    We have a separate tx signing flow for stake pool registration because it's a
+    transaction where the witnessable entries (i.e. inputs, withdrawals, etc.)
+    in the transaction are not supposed to be controlled by the HW wallet, which
+    means the user is vulnerable to unknowingly supplying a witness for an UTXO
+    or other tx entry they think is external, resulting in the co-signers
+    gaining control over their funds (Something SLIP-0019 is dealing with for
+    BTC but no similar standard is currently available for Cardano). Hence we
+    completely forbid witnessing inputs and other entries of the transaction
+    except the stake pool certificate itself and we provide a witness only to the
+    user's staking key in the list of pool owners.
+    """
+    try:
+        _validate_stake_pool_registration_tx_structure(msg)
+
+        _ensure_only_external_inputs(msg.inputs)
```

Comment:
in the above context, "supplying a witness for an UTXO they think is external" makes it seem that `_ensure_only_external_inputs` do the opposite of what it says :)

suggest renaming to `_ensure_no_signing_inputs` and adding a blurb about it to its docstring